src/nvim/cursor_shape.c

```diff
@@ -7,40 +7,108 @@
 #include "nvim/charset.h"
 #include "nvim/strings.h"
 #include "nvim/syntax.h"
+#include "nvim/api/private/helpers.h"
+#include "nvim/ui.h"
 
-/*
- * Handling of cursor and mouse pointer shapes in various modes.
- */
-
+/// Handling of cursor and mouse pointer shapes in various modes.
 static cursorentry_T shape_table[SHAPE_IDX_COUNT] =
 {
-  /* The values will be filled in from the 'guicursor' and 'mouseshape'
-   * defaults when Vim starts.
-   * Adjust the SHAPE_IDX_ defines when making changes! */
-  {0, 0, 0, 700L, 400L, 250L, 0, 0, "n", SHAPE_CURSOR+SHAPE_MOUSE},
-  {0, 0, 0, 700L, 400L, 250L, 0, 0, "v", SHAPE_CURSOR+SHAPE_MOUSE},
-  {0, 0, 0, 700L, 400L, 250L, 0, 0, "i", SHAPE_CURSOR+SHAPE_MOUSE},
-  {0, 0, 0, 700L, 400L, 250L, 0, 0, "r", SHAPE_CURSOR+SHAPE_MOUSE},
-  {0, 0, 0, 700L, 400L, 250L, 0, 0, "c", SHAPE_CURSOR+SHAPE_MOUSE},
-  {0, 0, 0, 700L, 400L, 250L, 0, 0, "ci", SHAPE_CURSOR+SHAPE_MOUSE},
-  {0, 0, 0, 700L, 400L, 250L, 0, 0, "cr", SHAPE_CURSOR+SHAPE_MOUSE},
-  {0, 0, 0, 700L, 400L, 250L, 0, 0, "o", SHAPE_CURSOR+SHAPE_MOUSE},
-  {0, 0, 0, 700L, 400L, 250L, 0, 0, "ve", SHAPE_CURSOR+SHAPE_MOUSE},
-  {0, 0, 0,   0L,   0L,   0L, 0, 0, "e", SHAPE_MOUSE},
-  {0, 0, 0,   0L,   0L,   0L, 0, 0, "s", SHAPE_MOUSE},
-  {0, 0, 0,   0L,   0L,   0L, 0, 0, "sd", SHAPE_MOUSE},
-  {0, 0, 0,   0L,   0L,   0L, 0, 0, "vs", SHAPE_MOUSE},
-  {0, 0, 0,   0L,   0L,   0L, 0, 0, "vd", SHAPE_MOUSE},
-  {0, 0, 0,   0L,   0L,   0L, 0, 0, "m", SHAPE_MOUSE},
-  {0, 0, 0,   0L,   0L,   0L, 0, 0, "ml", SHAPE_MOUSE},
-  {0, 0, 0, 100L, 100L, 100L, 0, 0, "sm", SHAPE_CURSOR},
```

Comment:
I didn't realize that the short strings "n", "v", etc. could be used to identify these entries. Probably those should be used (over RPC) instead of creating the long-form names.